tests/arch/arc/arc_vpx_lock/src/main.c

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/ztest.h>
+
+#include <zephyr/arch/arc/v2/dsp/arc_dsp.h>
+
+#ifndef CONFIG_ARC_DSP
+#error "Rebuild with the ARC_DSP config option enabled"
+#endif
+
+#ifndef CONFIG_ARC_VPX_COOPERATIVE_SHARING
+#error "Rebuild with the ARC_VPX_COOPERATIVE_SHARING config option enabled"
+#endif
+
+#define STACK_SIZE (1024 + CONFIG_TEST_EXTRA_STACK_SIZE)
+
+#define ISR_VPX_LOCK         0
+#define ISR_VPX_UNLOCK       1
+
+struct test_data {
+	k_tid_t thread;
+	uint32_t cpu_mask;
+};
+
+static void timer_func(struct k_timer *timer);
+
+K_THREAD_STACK_DEFINE(helper_stack, STACK_SIZE);
+
+static K_TIMER_DEFINE(my_timer, timer_func, NULL);
+static K_SEM_DEFINE(test_sem1, 0, 1);
+static K_SEM_DEFINE(test_sem2, 0, 1);
+
+static struct test_data data;
+static struct k_thread helper_thread;
+
+static volatile int isr_result;
+static volatile unsigned int isr_vpx_lock_id;
+
+/**
+ * Obtain the current CPU id.
+ */
+static int current_cpu_id_get(void)
+{
+	int key;
+	int id;
+
+	key = arch_irq_lock();
+	id = _current_cpu->id;
+	arch_irq_unlock(key);
+
+	return id;
+}
+
+static void test_helper_entry(void *p1, void *p2, void *p3)
+{
+	uint32_t i;
+
+	while (1) {
+		k_sem_take(&test_sem1, K_FOREVER);
+
+#if CONFIG_SCHED_CPU_MASK
+		/*
+		 * Suspend the target thread to ensure that it does not run
+		 * while we modify the set of CPUs upon which it may execute.
+		 */
+
+		k_thread_suspend(data->thread);
+
+		for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
+			if ((data->cpu_mask & BIT(i)) == 0) {
+				k_thread_cpu_mask_disable(data->thread, i);
+			} else {
+				k_thread_cpu_mask_enable(data->thread, i);
+			}
+		}
```

Comment:
Hmm, it looks like over-complicated way to do `k_thread_cpu_pin`.

---

Moreover the `test_helper_entry` itself looks over-complicated (or I'm missing some test logic here).
If we just need to pin a test thread to one CPU I'd just do it this way (and drop `test_helper_entry` entirely):

```
ZTEST(vpx_lock, test_case)
{
        k_thread_create(&test_payload, ...);
        k_thread_cpu_pin(&test_payload, cpu_id);
        k_thread_start(&test_payload);
}
```

Reply:
This is a leftover artefact of previous implementations and test versions when the cpu_mask check in arc_vpx_lock() was an if-statement and not an assert. I will simplify it.